### ISSUE TITLE
Installation on an raspi pi4: ModuleNotFoundError: No module named 'jsonschema' , Smarthome fhem

### ISSUE THREAD

**Issue by ollibaba** [mpepping/solarman-mqtt#40]
Hi,
I installed and configured the Solarman-Mqtt on an raspi and always got the error message below: 
/opt/solarman-mqtt $ python3 run.py -s -d -h   # this is for testing only single start
Traceback (most recent call last):
  File "run.py", line 5, in <module>
    import solarman
  File "/opt/solarman-mqtt/solarman/__init__.py", line 12, in <module>
    from .helpers import ConfigCheck, HashPassword
  File "/opt/solarman-mqtt/solarman/helpers.py", line 7, in <module>
    from jsonschema import validate
ModuleNotFoundError: **No module named 'jsonschema'**

the raspi has installed as standard python 2.7. and 3.7.

Then I changed in requirements.txt the option to jsonschema==2.6.0
and started again with standard python2:
 python run.py -s -d -h
Traceback (most recent call last):
  File "run.py", line 5, in <module>
    import solarman
  File "/opt/solarman-mqtt/solarman/__init__.py", line 11, in <module>
    from .api import SolarmanApi, ConstructData
  File "/opt/solarman-mqtt/solarman/api.py", line 45
    url = **f**"//account/v1.0/token?appId={appid}&language=en"

Here it seems that the f option in line 45 makes the problem because py2 isn´t able to handle this.

What should I do? Do you have an idea?

**Comment by mpepping** [comment]
Hi @ollibaba, the [README doesn't state it explicitly](https://github.com/mpepping/solarman-mqtt#using-python), but the app is Python 3 only. Python 3.7 should be good though:
https://github.com/mpepping/solarman-mqtt#using-python

**Comment by ollibaba** [comment]
Hi,
the problem is probably the result from:
 **sudo pip install -r requirements.txt**
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting paho_mqtt==1.5.1 (from -r requirements.txt (line 1))
  Downloading https://files.pythonhosted.org/packages/32/d3/6dcb8fd14746fcde6a556f932b5de8bea8fedcb85b3a092e0e986372c0e7/paho-mqtt-1.5.1.tar.gz (101kB)
    100%  102kB 1.6MB/s
Collecting jsonschema==4.4.0 (from -r requirements.txt (line 2))
  Could not find a version that satisfies the requirement jsonschema==4.4.0 (from -r requirements.txt (line 2)) (from versions: 0.1a0, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8.0, 1.0.0, 1.1.0, 1.2.0, 1.3.0, 2.0.0, 2.1.0, 2.2.0, 2.3.0, 2.4.0, 2.5.0, 2.5.1, 2.6.0, 3.0.0a1, 3.0.0a2, 3.0.0a3, 3.0.0a4, 3.0.0a5, 3.0.0a6, 3.0.0b1, 3.0.0b2, 3.0.0b3, 3.0.0, 3.0.1, 3.0.2, 3.1.0, 3.1.1, 3.2.0, 4.0.0a1, 4.0.0a2, 4.0.0a3, 4.0.0a4, 4.0.0a5, 4.0.0a6, 4.0.0)
**No matching distribution found for jsonschema==4.4.0 (from -r requirements.txt (line 2))**
pi@raspi4-1:/opt/solarman-mqtt $

and this ends in the error messages when calling **python3 run.py ...** afterwards.

**Comment by ollibaba** [comment]
Next try installed jsonschema 4.0.0 but see later:
sudo pip**3** install -r requirements.txt    # <-- the change to pip3 works

Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting paho_mqtt==1.5.1 (from -r requirements.txt (line 1))
  Downloading https://www.piwheels.org/simple/paho-mqtt/paho_mqtt-1.5.1-py3-none-any.whl (74kB)
    100% |████████████████████████████████| 81kB 834kB/s
Collecting jsonschema==4.0.0 (from -r requirements.txt (line 3))
  Downloading https://files.pythonhosted.org/packages/e0/d9/05587ac378b9fd2c352c6f024f13240168365bd753a7e8007522b7025267/jsonschema-4.0.0-py3-none-any.whl (69kB)
    100% |████████████████████████████████| 71kB 1.6MB/s
Collecting importlib-metadata; python_version < "3.8" (from jsonschema==4.0.0->-r requirements.txt (line 3))
  Downloading https://files.pythonhosted.org/packages/b5/64/ef29a63cf08f047bb7fb22ab0f1f774b87eed0bb46d067a5a524798a4af8/importlib_metadata-5.0.0-py3-none-any.whl
Collecting attrs>=17.4.0 (from jsonschema==4.0.0->-r requirements.txt (line 3))
  Downloading https://files.pythonhosted.org/packages/f2/bc/d817287d1aa01878af07c19505fafd1165cd6a119e9d0821ca1d1c20312d/attrs-22.1.0-py2.py3-none-any.whl (58kB)
    100% |████████████████████████████████| 61kB 2.0MB/s
Collecting pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0 (from jsonschema==4.0.0->-r requirements.txt (line 3))
  Downloading https://www.piwheels.org/simple/pyrsistent/pyrsistent-0.18.1-cp37-cp37m-linux_armv7l.whl (82kB)
    100% |████████████████████████████████| 92kB 936kB/s
Collecting typing-extensions>=3.6.4; python_version < "3.8" (from importlib-metadata; python_version < "3.8"->jsonschema==4.0.0->-r requirements.txt (line 3))
  Downloading https://files.pythonhosted.org/packages/0b/8e/f1a0a5a76cfef77e1eb6004cb49e5f8d72634da638420b9ea492ce8305e8/typing_extensions-4.4.0-py3-none-any.whl
Collecting zipp>=0.5 (from importlib-metadata; python_version < "3.8"->jsonschema==4.0.0->-r requirements.txt (line 3))
  Downloading https://files.pythonhosted.org/packages/40/8a/d63273ed0fa4a3d06f77e7b043f6577d8894e95515b0c187c52e2c0efabb/zipp-3.10.0-py3-none-any.whl
**Installing collected packages: paho-mqtt, typing-extensions, zipp, importlib-metadata, attrs, pyrsistent, jsonschema
Successfully installed attrs-22.1.0 importlib-metadata-5.0.0 **jsonschema-4.0.0** **paho-mqtt-1.5.1** pyrsistent-0.18.1 typing-extensions-4.4.0 zipp-3.10.0

**seems ok, but then several new errors:**

pi@raspi4-1:/opt/solarman-mqtt $ **sudo python3 run.py -s**
Traceback (most recent call last):
  File "run.py", line 14, in <module>
    main()
  File "run.py", line 10, in main
    solarman.main()
  File "/opt/solarman-mqtt/solarman/__init__.py", line 184, in main
    single_run(args.file)
  File "/opt/solarman-mqtt/solarman/__init__.py", line 56, in single_run
    pvdata = SolarmanApi(config)
  File "/opt/solarman-mqtt/solarman/api.py", line 26, in __init__
    self.station_realtime = self.get_station_realtime()
  File "/opt/solarman-mqtt/solarman/api.py", line 66, in get_station_realtime
    "Authorization": "bearer " + self.token,
TypeError: can only concatenate str (not "NoneType") to str
pi@raspi4-1:/opt/solarman-mqtt $

**Comment by ollibaba** [comment]
ok. now I tried   **sudo python3 run.py --validate** 
with the error result **inverterID** is no string. Set it in "xxx" and then it was ok. --> **is this ok?**
Then the 2nd error. It seems that my appID I got from IGEN is to long (16digits) tried it with 15digits (same id with only 15digits as test, know it´s wrong) and it works. No error more while validation the config. Asked now for a new/correct pair appID and secret from IGEN. 
the result from:  **sudo python3 run.py -s**
is now:
**ERROR:root:Unable to fetch token: 'access_token'**

**Comment by ollibaba** [comment]
@mpepping 
Hi Martijn,
would you pls. so kind and check my info above especially the last comment with the changes in the **config.json**.
I think the part with the jsonsschema ist clarified after using the **pip3** command for requirements installation.
I had to change **in the config the inverterID** to a string with "" due to the error message while validation and as 2nd pt it seems that the **appID is to long with 16 digits**. Could you check this both points and give me an response.
many thanks Oliver

**Comment by mpepping** [comment]
Hi @ollibaba, strange thing with the access key you have there! I would suggest to test the key using curl. An example can be found here: https://github.com/mpepping/solarman-mqtt#how-to-get-all-required-input-for-the-config-file

That's probably the best way to determine if the app key that you received works. Because, the `ERROR:root:Unable to fetch token: 'access_token'` error means that the app key used is not correct.

Btw, wrapping the `inverterID` in double quotes (your `"xxx"` example) is the right way to make a string from that number.

**Comment by ollibaba** [comment]
Thanks. yes very strange.I think the id and key doesn't work and ordered a new one from IGEN. Tommorow I can try the curl command. Will give an response then. Greetings Oliver gesendet von mobileAm 30.10.2022 19:50 schrieb Martijn Pepping ***@***.***>:
Hi @ollibaba, strange thing with the access key you have there! I would suggest to test the key using curl. An example can be found here: https://github.com/mpepping/solarman-mqtt#how-to-get-all-required-input-for-the-config-file
That's probably the best way to determine if the app key that you received works. Because, the ERROR:root:Unable to fetch token: 'access_token' error means that the app key used is not correct.
Btw, wrapping the inverterID in double quotes (your "xxx" example) is the right way to make a string from that number.

**Comment by ollibaba** [comment]
@mpepping 
Hi Martijn,
**the result with the curl  cmd line:**
curl --location --request POST 'https://api.solarmanpv.com//account/v1.0/token?appId=**my16digitsid**&language=en' \
  --header 'Content-Type: application/json' \
  --data-raw '{
  "appSecret": "**the32digitssecret**",
  "email": "mymail_for_the_app@xyz.de",
  "password": "**pashhash: result from run.py --pashhash...cmd**"
}'

**is:**

{"code":"2101025","msg":"{\"error\":\"**invalid_request**\",\"error_description\":\"\",\"code\":\"**AUTH_INVALID_USERNAME_OR_PASSWORD**\",\"param\":null}","success":false,"requestId":"9e8adb634b3a4db9","access_token":null,"token_type":null,"refresh_token":null,"expires_in":null,"scope":null,"uid":null}

Now I´m still waiting for a new pair of appid +secret from IGEN. 
Hopefully it will work then. 
Greetings Oliver

**Comment by ollibaba** [comment]
@mpepping 
Hi Martijn,
the best thing at first it´s now running with a little problem. But **to this point later**.
The error messages were all ambiguous.
The main mistake from my side was the wrong passhash in the config.
I generated it with ...**run.py --create-passhash CREATE_PASSHASH** 
Now after an inspiration :-) I know that for the word: **CREATE_PASSHASH** I have to input **the real password from the (Android) App Login**. Yes silly I know! mea culpa.
**Ok**. The **2nd thing in the config** was I had to set the value of the **stationId in the config in double quotes (")** due to an error message while starting the run.py. 
It´s no problem that the appID I got from IGEN service has 16digits. This is allowed and correct.
**That´s in principle all.**
in conclusion for the **raspi and rasbian OS** you have to install the requirements with an **pip3** command and not pip (!):
**sudo pip3 install -r requirements.txt**
Then adapt the config.json as described above and all is OK.
**But now the little problem in the end**
at the moment I´m importing the mqtt messages in my smarthome on raspi **fhem** and every 5 minutes (standard) I see the messages. But every time with a new device. For example: MQTT2_solarmanpv_mqtt_168, 5min later MQTT2_solarmanpv_mqtt_234, then after 5min MQTT2_solarmanpv_mqtt_297....969 endless until inverter is offline  due to nighttime shutdown.
**Do you have an idea to stop this behaviar?**
Greetings Oliver

**Comment by ollibaba** [comment]
@mpepping 
Hi Martijn,
me again. I thing the solution for the problem above is to change an entry in the **mqtt.py** in **line 32**
**from:** client_id = f"solarmanpv-mqtt-{random.randint(0, 1000)}"   # <-- remark: I don´t know the reason of this syntax 
**to:**  client_id = f"solarmanpv-mqtt-**nomatter**"
and then all mqtt messages will be send  to solarmanpv-mqtt-**nomatter**"
**That´s it !**
So let´s see tommorow at daylight times if it works over the day...
Greetings Oliver
* and many thanks for the scripts. Perhaps you can make a little update with my some of my remarks in the readme

**Comment by ollibaba** [comment]
Additionally a remark:
for the raspi I use the syntax: **sudo python3 run.py -s**     <----always single run
and **updated the crontab with a line:  
*/5 6-18 * * * sudo python3 /opt/solarman-mqtt/run.py -s -f /opt/solarman-mqtt/config.json 
so you will not load the IGEN solarman Server while nighttimes
Docker scripts you provided here will not work with the pi due to hardware environment restrictions which cannot be loaded while installing the docker version.
:-) Oliver
